### DISCUSSION TITLE
How to mip install from gitlab repository?

### DISCUSSION THREAD

**Discussion by OlivierLenoir** [micropython/micropython#13781] [General]
How to _mip install_ from _gitlab.com_ repository?

I've created a test repository on [gitlab.com](https://gitlab.com/olivierlenoir/MipTest).

- Is my _package.json_ file correct?
- Can we _mpremote mip install_ from _gitlab_ and how?

**Comment by jimmo** [comment]
This should work -- `mip` supports installing from any HTTP(s) source.

However, `mip` has a special convenience shortcut for GitHub that allows you to write `github:org/repo` to access the files.

If you want to use `gitlab.com:olivierlenoir/...` like in your example `package.json` you would need to add support for GitLab URLs to `mip`. See https://github.com/micropython/micropython-lib/blob/master/micropython/mip/mip/__init__.py#L64 for where the GitHub support is implemented.

**Reply by OlivierLenoir** [reply]
I don't know how to test an implementation of _gitlab:_, but here is how we could implement it (to be tested):
[https://github.com/micropython/micropython-lib/blob/master/micropython/mip/mip/__init__.py#L64](https://github.com/micropython/micropython-lib/blob/master/micropython/mip/mip/__init__.py#L64)
```python
def _rewrite_url(url, branch=None):
    if not branch:
        branch = "HEAD"
    if url.startswith("github:"):
        url = url[7:].split("/")
        url = (
            "https://raw.githubusercontent.com/"
            + url[0]
            + "/"
            + url[1]
            + "/"
            + branch
            + "/"
            + "/".join(url[2:])
        )
    if url.startswith("gitlab:"):
        url = url[7:].split("/")
        url = (
            "https://gitlab.com/"
            + url[0]
            + "/"
            + url[1]
            + "/-/raw/"
            + branch
            + "/"
            + "/".join(url[2:])
        )
    return url

#...


def _install_package(package, index, target, version, mpy):
    if (
        package.startswith("http://")
        or package.startswith("https://")
        or package.startswith("github:")
        or package.startswith("gitlab:")
    ):
#...
``` 
Could you check if it can be implement and documented?

**Reply by OlivierLenoir** [reply]
@jimmo, I've been able to test my above proposition to implement _gitlab:_ by modifying my local _mip.py_.
It works.
```shell
$ mpremote mip install gitlab:olivierlenoir/MipTest
Install gitlab:olivierlenoir/MipTest
Installing gitlab:olivierlenoir/MipTest/package.json to /lib
Installing: /lib/logger.py
Installing: /lib/softapwebserver.py         
Done                                        
``` 
My _gitlab.com_ repository [https://gitlab.com/olivierlenoir/MipTest](https://gitlab.com/olivierlenoir/MipTest/-/commit/2ebdd18409072e4019dfdaeb1a2ce847749ffd98)

**Reply by OlivierLenoir** [reply]
@jimmo, I've pull request to implement _gitlab:_.

**Reply by OlivierLenoir** [reply]
Both pull requests are link to the same implementation of mip install from GitLab.

- https://github.com/micropython/micropython-lib/pull/816
- https://github.com/micropython/micropython/pull/13791

Signed-off-by: Olivier Lenoir olivier.len02@gmail.com

-------------------------------------------------------
https://github.com/micropython/micropython-lib/pull/816

Modify _rewrite_url() to allow mip install from gitlab: repository.

Signed-off-by: Olivier Lenoir olivier.len02@gmail.com

-----------------------------------------------------
https://github.com/micropython/micropython/pull/13791

mpremote: Implement mip install from GitLab.

tools/mpremote/README.md: Add mip install gitlab:org/repo@branch.
Add example mpremote mip install gitlab:org/repo@branch.

tools/mpremote/mpremote/mip.py: Implement install from GitLab.
Modify _rewrite_url() to allow mip install from gitlab: repository.

mpremote/mpremote/main.py: Add gitlab:org/repo, gitlab:org/repo@branch.
Modify argparse_mip() for gitlab:org/repo, gitlab:org/repo@branch.

docs/reference/packages.rst: Mip install from GitLab.
Describe how to mip install from GitLab.

docs/reference/mpremote.rst: Install package from GitLab.
Install the package from the specified branch at org/repo on GitLab.

Signed-off-by: Olivier Lenoir olivier.len02@gmail.com

**Comment by OlivierLenoir** [comment]
I didn't find the equivalent of _https://raw.githubusercontent.com/_ for _gitlab_, but here is what I did.

```shell
$ mpremote mip install https://gitlab.com/olivierlenoir/MipTest/-/raw/main
Install https://gitlab.com/olivierlenoir/MipTest/-/raw/main
Installing https://gitlab.com/olivierlenoir/MipTest/-/raw/main/package.json to /lib
Installing: /lib/logger.py
Installing: /lib/softapwebserver.py         
Done                                        
``` 
_package.json_
```
{
    "urls": [
        ["logger.py", "https://gitlab.com/olivierlenoir/MipTest/-/raw/main/micropython/logger.py"],
        ["softapwebserver.py", "https://gitlab.com/olivierlenoir/MipTest/-/raw/main/micropython/softapwebserver.py"]
    ],
    "deps": [
    ],
    "version": "0.1.0"
}
```

It works.
Thanks,
